pyadjoint/tape.py

```diff
@@ -319,40 +417,59 @@ def optimize(self, controls=None, functionals=None):
     def optimize_for_controls(self, controls):
         # TODO: Consider if we want Enlist wherever it is possible. Like in this case.
         # TODO: Consider warning/message on empty tape.
-        blocks = self.get_blocks()
         nodes = set([control.block_variable for control in controls])
-        valid_blocks = []
-
-        for block in blocks:
-            depends_on_control = False
-            for dep in block.get_dependencies():
-                if dep in nodes:
-                    depends_on_control = True
-
-            if depends_on_control:
-                for output in block.get_outputs():
-                    if output in nodes:
-                        raise RuntimeError("Control depends on another control.")
-                    nodes.add(output)
-                valid_blocks.append(block)
-        self._blocks = valid_blocks
+        discarded_variables = set()
+        optimized_timesteps = TimeStepSequence()
+
+        for step in self._blocks.steps:
+            optimized_timesteps.append_step()
+
+            for block in step:
+                depends_on_control = False
+                for dep in block.get_dependencies():
+                    if dep in nodes:
+                        depends_on_control = True
```

Comment:
```suggestion
                        depends_on_control = True
                        break
```
I would suggest adding a break here, as we shouldn't need to continue looping through dependencies if we have found a dependency in an earlier node/